File: web/src/lib/geo.ts
import type { Coordinate } from "./types";

export const GAZETTEER: Record<string, Coordinate> = {
  київ: { lat: 50.45, lng: 30.52 },
  киев: { lat: 50.45, lng: 30.52 },
  харків: { lat: 49.99, lng: 36.23 },
  харьков: { lat: 49.99, lng: 36.23 },
  харківщина: { lat: 49.95, lng: 36.37 },
  харьковщина: { lat: 49.95, lng: 36.37 },
  дніпро: { lat: 48.46, lng: 35.05 },
  днепр: { lat: 48.46, lng: 35.05 },
  одеса: { lat: 46.48, lng: 30.73 },
  одесса: { lat: 46.48, lng: 30.73 },
  полтава: { lat: 49.59, lng: 34.55 },
  кременчук: { lat: 49.07, lng: 33.42 },
  запоріжжя: { lat: 47.84, lng: 35.14 },
  запорожье: { lat: 47.84, lng: 35.14 },
  суми: { lat: 50.91, lng: 34.8 },
  чернігів: { lat: 51.49, lng: 31.29 },
  чернигов: { lat: 51.49, lng: 31.29 },
  львів: { lat: 49.84, lng: 24.03 },
  львов: { lat: 49.84, lng: 24.03 },
  миколаїв: { lat: 46.98, lng: 31.99 },
  николаев: { lat: 46.98, lng: 31.99 },
  чугуїв: { lat: 49.84, lng: 36.69 },
  чугуев: { lat: 49.84, lng: 36.69 },
  ізюм: { lat: 49.21, lng: 37.27 },
  изюм: { lat: 49.21, lng: 37.27 },
  купянськ: { lat: 49.71, lng: 37.62 },
  купянск: { lat: 49.71, lng: 37.62 },
};

export const LOCATION_ALIASES: Record<string, string> = {
  харкові: "харків",
  харькове: "харьков",
  харківського: "харків",
  харьковского: "харьков",
  харківщину: "харківщина",
  харьковщину: "харьковщина",
};

export const MAJOR_TARGETS = [
  "київ",
  "харків",
  "дніпро",
  "одеса",
  "полтава",
  "запоріжжя",
  "суми",
  "чернігів",
  "львів",
  "миколаїв",
];

export function normalizeText(value: string): string {
  return value.trim().toLowerCase().replace(/ё/g, "е");
}

export function lookupCoordinate(location: string | null): Coordinate | null {
  if (!location) return null;
  const normalized = normalizeText(location);
  const canonical = LOCATION_ALIASES[normalized] ?? normalized;
  return GAZETTEER[canonical] ?? null;
}

export function destinationPoint(
  origin: Coordinate,
  bearingDeg: number,
  distanceKm: number,
): Coordinate {
  const radiusKm = 6371;
  const bearing = (bearingDeg * Math.PI) / 180;
  const lat1 = (origin.lat * Math.PI) / 180;
  const lng1 = (origin.lng * Math.PI) / 180;
  const angular = distanceKm / radiusKm;

  const lat2 = Math.asin(
    Math.sin(lat1) * Math.cos(angular) +
      Math.cos(lat1) * Math.sin(angular) * Math.cos(bearing),
  );
  const lng2 =
    lng1 +
    Math.atan2(
      Math.sin(bearing) * Math.sin(angular) * Math.cos(lat1),
      Math.cos(angular) - Math.sin(lat1) * Math.sin(lat2),
    );

  return {
    lat: (lat2 * 180) / Math.PI,
    lng: (lng2 * 180) / Math.PI,
  };
}

export function haversineKm(a: Coordinate, b: Coordinate): number {
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const lat1 = toRad(a.lat);
  const lat2 = toRad(b.lat);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;
  return 6371 * 2 * Math.atan2(Math.sqrt(h), Math.sqrt(1 - h));
}

export function interpolatePath(path: Coordinate[], progress: number): Coordinate | null {
  if (!path.length) return null;
  if (path.length === 1) return path[0];
  const clamped = Math.max(0, Math.min(1, progress));
  const total = path.length - 1;
  const scaled = clamped * total;
  const idx = Math.floor(scaled);
  const t = scaled - idx;
  if (idx >= total) return path[path.length - 1];
  const a = path[idx];
  const b = path[idx + 1];
  return {
    lat: a.lat + (b.lat - a.lat) * t,
    lng: a.lng + (b.lng - a.lng) * t,
  };
}

File: backend/app/geo.py
from __future__ import annotations

import math

from app.schemas import Coordinate, Direction, ThreatType


# Minimal MVP gazetteer. Replace with a complete geocoder/PostGIS table before operational use.
GAZETTEER: dict[str, Coordinate] = {
    "миколаївщина": Coordinate(latitude=47.16, longitude=31.98),
    "николаевщина": Coordinate(latitude=47.16, longitude=31.98),
    "миколаїв": Coordinate(latitude=46.98, longitude=31.99),
    "николаев": Coordinate(latitude=46.98, longitude=31.99),
    "одеса": Coordinate(latitude=46.48, longitude=30.73),
    "одесса": Coordinate(latitude=46.48, longitude=30.73),
    "київ": Coordinate(latitude=50.45, longitude=30.52),
    "киев": Coordinate(latitude=50.45, longitude=30.52),
    "харків": Coordinate(latitude=49.99, longitude=36.23),
    "харьков": Coordinate(latitude=49.99, longitude=36.23),
    "харківщина": Coordinate(latitude=49.95, longitude=36.37),
    "харьковщина": Coordinate(latitude=49.95, longitude=36.37),
    "харківська область": Coordinate(latitude=49.95, longitude=36.37),
    "харьковская область": Coordinate(latitude=49.95, longitude=36.37),
    "чугуїв": Coordinate(latitude=49.84, longitude=36.69),
    "чугуев": Coordinate(latitude=49.84, longitude=36.69),
    "ізюм": Coordinate(latitude=49.21, longitude=37.27),
    "изюм": Coordinate(latitude=49.21, longitude=37.27),
    "купянськ": Coordinate(latitude=49.71, longitude=37.62),
    "купянск": Coordinate(latitude=49.71, longitude=37.62),
    "балаклія": Coordinate(latitude=49.46, longitude=36.84),
    "балаклея": Coordinate(latitude=49.46, longitude=36.84),
    "дергачі": Coordinate(latitude=50.11, longitude=36.12),
    "дергачи": Coordinate(latitude=50.11, longitude=36.12),
    "лозова": Coordinate(latitude=48.90, longitude=36.31),
    "дніпро": Coordinate(latitude=48.46, longitude=35.05),
    "днепр": Coordinate(latitude=48.46, longitude=35.05),
    "запоріжжя": Coordinate(latitude=47.84, longitude=35.14),
    "запорожье": Coordinate(latitude=47.84, longitude=35.14),
    "полтава": Coordinate(latitude=49.59, longitude=34.55),
    "кременчук": Coordinate(latitude=49.07, longitude=33.42),
    "черкаси": Coordinate(latitude=49.44, longitude=32.06),
    "суми": Coordinate(latitude=50.91, longitude=34.80),
    "чернігів": Coordinate(latitude=51.49, longitude=31.29),
    "чернигов": Coordinate(latitude=51.49, longitude=31.29),
    "львів": Coordinate(latitude=49.84, longitude=24.03),
    "львов": Coordinate(latitude=49.84, longitude=24.03),
}


LOCATION_ALIASES: dict[str, str] = {
    "миколаївщину": "миколаївщина",
    "николаевщину": "николаевщина",
    "харківського": "харків",
    "харьковского": "харьков",
    "харківщину": "харківщина",
    "харьковщину": "харьковщина",
    "харківщині": "харківщина",
    "харьковщине": "харьковщина",
    "харкові": "харків",
    "харькове": "харьков",
    "у харкові": "харків",
    "в харькове": "харьков",
    "полтавщину": "полтава",
}


DIRECTION_BEARINGS: dict[Direction, float] = {
    Direction.NORTH: 0,
    Direction.NORTHEAST: 45,
    Direction.EAST: 90,
    Direction.SOUTHEAST: 135,
    Direction.SOUTH: 180,
    Direction.SOUTHWEST: 225,
    Direction.WEST: 270,
    Direction.NORTHWEST: 315,
}


def normalize_location(text: str) -> str:
    return text.strip().lower().replace("ё", "е")


def lookup_coordinate(location: str) -> Coordinate | None:
    normalized = normalize_location(location)
    canonical = LOCATION_ALIASES.get(normalized, normalized)
    return GAZETTEER.get(canonical)


def risk_radius_km(threat_type: ThreatType, confidence: float) -> float:
    base_radius = {
        ThreatType.SHAHED: 35.0,
        ThreatType.UAV: 25.0,
        ThreatType.GERBERA: 20.0,
        ThreatType.KAB: 18.0,
        ThreatType.FPV: 8.0,
        ThreatType.RECON_DRONE: 22.0,
        ThreatType.MISSILE: 45.0,
        ThreatType.UNKNOWN: 30.0,
    }[threat_type]
    uncertainty_multiplier = 1.3 - min(max(confidence, 0.0), 1.0) * 0.4
    return round(base_radius * uncertainty_multiplier, 1)


def destination_point(origin: Coordinate, bearing_degrees: float, distance_km: float) -> Coordinate:
    radius_km = 6371.0
    bearing = math.radians(bearing_degrees)
    lat1 = math.radians(origin.latitude)
    lon1 = math.radians(origin.longitude)
    angular_distance = distance_km / radius_km

    lat2 = math.asin(
        math.sin(lat1) * math.cos(angular_distance)
        + math.cos(lat1) * math.sin(angular_distance) * math.cos(bearing)
    )
    lon2 = lon1 + math.atan2(
        math.sin(bearing) * math.sin(angular_distance) * math.cos(lat1),
        math.cos(angular_distance) - math.sin(lat1) * math.sin(lat2),
    )
    return Coordinate(latitude=math.degrees(lat2), longitude=math.degrees(lon2))


def risk_geometry(
    coordinate: Coordinate | None,
    direction: Direction,
    radius_km: float,
) -> dict:
    if coordinate is None:
        return {"type": "GeometryCollection", "geometries": []}

    if direction in DIRECTION_BEARINGS:
        end = destination_point(coordinate, DIRECTION_BEARINGS[direction], max(radius_km * 2.5, 25.0))
        return {
            "type": "LineString",
            "coordinates": [
                [coordinate.longitude, coordinate.latitude],
                [end.longitude, end.latitude],
            ],
        }

    return {
        "type": "Point",
        "coordinates": [coordinate.longitude, coordinate.latitude],
    }
